### PR TITLE
Issue 1205: BigQuery support for CREATE table OPTIONS ( description = 'desc' )

### LINKED ISSUE #1205: BigQuery: CREATE [OR REPLACE] table OPTIONS ( description = 'desc' ) is not supported
## Expected Behaviour

in our big query process, we have a lot of following queries:

```
CREATE OR REPLACE TABLE foo
OPTIONS
  (description = 'copy of bar') AS (
    SELECT * from bar
    )
```
## Observed Behaviour

Following error
```
Found unparsable section: 'CREATE OR REPLACE TABLE foo\nOPTIONS\n (d...'
```

## Steps to Reproduce

http://online.sqlfluff.com/fluffed?sql=H4sIAELg7WAC_yWKOwqAMBAFeyF3eJ2fO1jEsIUQTEhyAX8BC12JNt5eg9XAzChHMhCMgyOrpSIE2WlCZBaFsaE3gxcFUC3rNaftvDc-0KKc-XzAEdOYyhrSo8oX4EmTCmgQE--5_rr-8AL24jT3awAAAA%3D%3D&dialect=bigquery

### PR DESCRIPTION
**Pull request by barrywhart**
Resolves #1205

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_bigquery.py [hunk 1]
```diff
@@ -561,3 +562,86 @@ class SetStatementSegment(BaseSegment):
...
+        ),
+    )
+
+
+@bigquery_dialect.segment(replace=True)
+class CreateTableStatementSegment(BaseSegment):
+    """A `CREATE TABLE` statement."""
+
+    type = "create_table_statement"
+    # https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#create_table_statement
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_bigquery.py:617]: In addition to the "OPTIONS" clause, BigQuery "CREATE TABLE" also has "PARTITION BY" and "CLUSTER BY". Together, these differences seem big enough to justify replacing `CreateTableStatementSegment` rather than just customizing/extending it. That's why I went ahead and added support for "PARTITION BY" and "CLUSTER BY" even though they aren't part of issue #1205.

**Comment by barrywhart** [comment]
This is ready for review! It's a pretty straightforward PR. It's a bit larger than I expected because BigQuery has some other unusual "CREATE TABLE" options. Nothing tricky, though. @dmateusp @NiallRees

File: test/fixtures/parser/bigquery/create_table_columns_partition_options.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE newtable (
+    x INT64,
+    y INT64
+)
+PARTITION BY x, y
+CLUSTER BY x, y
+OPTIONS(description="foo")
```
Comment (dmohns) [inline test/fixtures/parser/bigquery/create_table_columns_partition_options.sql:7]: Small nit: This is not a valid BigQuery query

```
Only a single PARTITION BY expression is supported but found 2
```

Maybe you could replace it with something valid? For example a something that is probably commonly used

```suggestion
CREATE TABLE newtable (
    x TIMESTAMP,
    y INT64
)
PARTITION BY DATE(x)
CLUSTER BY x, y
OPTIONS(description="foo")
```

Comment (barrywhart) [reply test/fixtures/parser/bigquery/create_table_columns_partition_options.sql:7]: Good point -- I'll change this. Note that in general, SQLFluff dialects are often more permissive than the actual implementation, but that's usually for pragmatic reasons (speed or simplicity of implementation). The correct thing should be easy here, though. 👍🏽

File: test/fixtures/parser/ansi/truncate_a.yml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 file:
   statement:
     truncate_table:
-    - keyword: truncate
-    - table_reference:
+      keyword: truncate
+      table_reference:
```
Comment (tunetheweb) [inline test/fixtures/parser/ansi/truncate_a.yml:5]: Nit, but is it confusing to make this unrelated change in this PR?

File: test/fixtures/parser/snowflake/snowflake_escape.yml [hunk 4]
```diff
@@ -2,15 +2,15 @@ file:
...
-            literal: "'c\\' '"
-            alias_expression:
-              keyword: as
-              identifier: escaped
-        - comma: ','
-        - select_clause_element:
-            literal: "'c\\' '' '"
-            alias_expression:
-              keyword: as
-              identifier: escaped_double
```
Comment (tunetheweb) [inline test/fixtures/parser/snowflake/snowflake_escape.yml:16]: Again, is it confusing to add these unrelated changes to this PR?

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Test it on a few of my `CREATE` statements (we tend not to use `OPTIONS`, but do use `PARTITION BY` and `CLUSTER BY`) and think I've identified a small improvement.

Plus some other nits.

But over all the approach looks good to me!

File: src/sqlfluff/dialects/dialect_bigquery.py [hunk 5]
```diff
@@ -604,3 +605,96 @@ class SetStatementSegment(BaseSegment):
...
+
+
+@bigquery_dialect.segment()
+class PartitionBySegment(BaseSegment):
+    """PARTITION BY partition_expression."""
+
+    type = "partition_by_segment"
+    match_grammar = StartsWith(
+        "PARTITION",
+        terminator=OneOf("CLUSTER", "OPTIONS", Ref("DelimiterSegment")),
```
[the remark this answers is not in the document]
Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_bigquery.py:617]: Great catch! Making this change now and adding a test.

File: src/sqlfluff/dialects/dialect_bigquery.py [hunk 6]
```diff
@@ -604,3 +605,96 @@ class SetStatementSegment(BaseSegment):
...
+        Ref("ExpressionSegment"),
+    )
+
+
+@bigquery_dialect.segment()
+class ClusterBySegment(BaseSegment):
+    """CLUSTER BY clustering_column_list."""
+
+    type = "cluster_by_segment"
+    # PARTITION BY partition_expression
```
[the remark this answers is not in the document]
Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_bigquery.py:632]: 👍 WIll do!

File: src/sqlfluff/dialects/dialect_bigquery.py [hunk 7]
```diff
@@ -604,3 +605,96 @@ class SetStatementSegment(BaseSegment):
...
+    match_grammar = StartsWith(
+        "CLUSTER",
+        terminator=OneOf("OPTIONS", Ref("DelimiterSegment")),
```
[the remark this answers is not in the document]
Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_bigquery.py:635]: 👍

File: test/fixtures/parser/ansi/truncate_a.yml (hunk 3 above)
Comment (barrywhart) [reply test/fixtures/parser/ansi/truncate_a.yml:5]: This is a cosmetic and unintentional but automated change that comes up occasionally when running `test/generate_parse_fixture_yml.py`. We've tried to set the YAML formatting options in that script so the output is always consistent, but have not yet succeeded in avoiding all such changes. Definitely a nuisance! Got any ideas about how to fix that script? I haven't looked at it in a few months.

File: test/fixtures/parser/bigquery/create_or_replace_table_options_as.sql [hunk 8]
```diff
@@ -0,0 +1,5 @@
+CREATE OR REPLACE TABLE foo
+OPTIONS
+  (description = 'copy of bar') AS (
+    SELECT * from bar
+    )
```
[the remark this answers is not in the document]
Comment (barrywhart) [reply test/fixtures/parser/bigquery/create_or_replace_table_options_as.sql:5]: 👍

File: test/fixtures/parser/bigquery/create_table_options_as.sql [hunk 9]
```diff
@@ -0,0 +1,3 @@
+CREATE OR REPLACE TABLE foo
+OPTIONS (description = 'copy of bar')
+AS (SELECT * from bar)
```
[the remark this answers is not in the document]
Comment (barrywhart) [reply test/fixtures/parser/bigquery/create_table_options_as.sql:3]: 👍

File: test/fixtures/parser/snowflake/snowflake_escape.yml (hunk 4 above)
Comment (barrywhart) [reply test/fixtures/parser/snowflake/snowflake_escape.yml:16]: This is another of what I described above -- unintentional changes introduced by the script. A nuisance for sure, but the right fix is to fix the script so the output is consistently formatted.

File: test/fixtures/parser/ansi/truncate_a.yml (hunk 3 above)
Comment (tunetheweb) [reply test/fixtures/parser/ansi/truncate_a.yml:5]: Yeah that's fine. Wasn't sure what convention was here: keep changes solely related to this PR, or just accept automatic formatting changes here (both options have their merits!), so thought I'd raise.

**Review by tunetheweb — APPROVED** [review]
LGTM!

**Review by NiallRees — APPROVED** [review]
